### PR TITLE
feat: include OpenShift container build and deploy steps

### PR DESCRIPTION
**Pull request by johanschvn**
How to build and deploy loopback4-example-shopping on OpenShift Container Platform

### CODE REVIEW

**Comment by raymondfeng** [comment]
@johanschvn Thank you for the contribution!

**Comment by mrmodise** [comment]
Thanks @johanschvn just 1 comment on my side. When dealing with containers and node i think you want to restrict the [container memory and CPU ](https://phoenixnap.com/kb/docker-memory-and-cpu-limit)as well as [max-old-space-size](https://stackoverflow.com/questions/48387040/node-js-recommended-max-old-space-size)

The above will ensure better performance of your app and infrastructure

**Comment by mrmodise** [comment]
@johanschvn this is really good. Looking forward to landing it

File: openshift/Dockerfile.recommender
Comment (mrmodise) [inline openshift/Dockerfile.recommender:12]: Feel free to ignore. Let's add --max-old-space-size perhaps restrict to 1GB or so. 512 which is the default is very small IMO

File: openshift/Dockerfile.shopping
Comment (mrmodise) [inline openshift/Dockerfile.shopping:12]: Same comment on --max-old-space-size

File: openshift/README.md
Comment (mrmodise) [inline openshift/README.md:10]: do they have to be run in order? if yes maybe mention the same

File: openshift/Dockerfile.recommender
Comment (johanschvn) [reply openshift/Dockerfile.recommender:12]: Not done in the Dockerfile.* in the root directory.
I do not want to add something that is not OpenShift related in this PR.

File: openshift/README.md
Comment (johanschvn) [reply openshift/README.md:10]: I think most people will try to run them in the order they are written and that will work.
I think it will add very little value explaining what commands can be run in a different order than listed.

Comment (mrmodise) [reply openshift/README.md:10]: I think we missed each other the recommendation was to mention that the comments should be executed in order. Remember not everyone who tries this out will be experienced in OpenShift. Change the line to:

```
First run the following commands in order in loopback4-example-shopping/openshift to create a new
```

File: openshift/Dockerfile.recommender
Comment (mrmodise) [reply openshift/Dockerfile.recommender:12]: If thats the case I recommend you reference the root yaml than duplicate. What do you think?

Comment (johanschvn) [reply openshift/Dockerfile.recommender:12]: The Dockerfile.* in root did not work for me in OpenShift.

Comment (mrmodise) [reply openshift/Dockerfile.recommender:12]: Do you mind elaborating on this point?

Comment (johanschvn) [reply openshift/Dockerfile.recommender:12]: Build terminates with this error if using ./Dockerfile.monorepo instead of openshift/Dockerfile.monorepo in base.yaml:
`error determining run uid: user: unknown user error looking up user "node"`

File: openshift/Dockerfile.shopping
Comment (johanschvn) [reply openshift/Dockerfile.shopping:12]: I do not think you have to hard code --max-old-space-size in the image. If you want to specify --max-old-space-size, I think you can do it using the NODE_OPTIONS environment variable.

File: openshift/Dockerfile.recommender
Comment (mrmodise) [reply openshift/Dockerfile.recommender:12]: Okay the users were different that means you probably running as root. Its discouraged but I know these Cloud platforms usually limit us

File: openshift/Dockerfile.shopping
Comment (mrmodise) [reply openshift/Dockerfile.shopping:12]: Thanks, its actually possible. You will change your CMD command to the below:

RUN node --max-old-space-size=79000 .

You can even do more in the case of this project where we are using `dotenv` by passing the dotenv bootstrapping command like below

RUN node -r dotenv/config --max-old-space-size=1024 .

With the above being said, I think you need to update the last line to accommodate `dotenv`

**Review by mrmodise — CHANGES_REQUESTED** [review]
For executing the CMD command on the Dockerfile.shopping you need accomodate the dotenv configuration by updating the CMD command to: 

CMD node -r dotenv/config --max-old-space-size=1024 .

File: openshift/Dockerfile.recommender
Comment (johanschvn) [reply openshift/Dockerfile.recommender:12]: OpenShift by default runs containers with random userid for security reasons, not root.
Why do you think it is run as root?

File: openshift/Dockerfile.shopping
Comment (johanschvn) [reply openshift/Dockerfile.shopping:12]: Last paragraph in my new README.md shows how to easily set environment variables in OpenShift.
The application appears to run fine without setting max-old-space-size. Why do you want to set it?
If you have some specific use case where it needs to be set you can do by setting environment variable NODE_OPTIONS to --max-old-space-size=79000 or whatever. I do not understand why you want it set when running in OpenShift, when it is not done otherwise!?

**Comment by johanschvn** [comment]
@mrmodise, I do not agree that Dockerfile.shopping needs to be changed.
Are you running in OpenShift and experience errors when not using option --max-old-space-size=1024 ?

File: openshift/Dockerfile.recommender
Comment (johanschvn) [reply openshift/Dockerfile.recommender:12]: Here is a blog post about userid in OpenShift:
https://cookbook.openshift.org/users-and-role-based-access-control/why-do-my-applications-run-as-a-random-user-id.html

File: openshift/Dockerfile.shopping
Comment (mrmodise) [reply openshift/Dockerfile.shopping:12]: Thanks this is for supporting bootstrapping dotenv. You can see more information on this: https://www.npmjs.com/package/dotenv

**Comment by mrmodise** [comment]
> @mrmodise, I do not agree that Dockerfile.shopping needs to be changed.
> Are you running in OpenShift and experience errors when not using option --max-old-space-size=1024 ?

Thanks however not experiencing errors does not mean we should not encourage best practices, as such, I recommend the change be applied before we land the PR.

**Comment by johanschvn** [comment]
"-r dotenv/config" is not needed, environment variables are set by the `oc set env` command if you use my instructions.
If you want to start node with --max-old-space-size=1024, I think you can do so with `oc set env dc/shopping NODE_OPTIONS=--max-old-space-size=1024`, but I have not tested; and I am not sure how to test it.
Documentation for NODE_OPTIONS are here:
https://nodejs.org/docs/latest-v12.x/api/cli.html#cli_node_options_options

File: openshift/recommender.yaml
Comment (mrmodise) [inline openshift/recommender.yaml:14]: Please also apply memory and cpu usage limits at container level to avoid unnecessary wastage. See for reference: https://docs.openshift.com/container-platform/3.11/admin_guide/limits.html#overview

**Review by mrmodise — CHANGES_REQUESTED** [review]

**Comment by mrmodise** [comment]
> "-r dotenv/config" is not needed, environment variables are set by the `oc set env` command if you use my instructions.
> If you want to start node with --max-old-space-size=1024, I think you can do so with `oc set env dc/shopping NODE_OPTIONS=--max-old-space-size=1024`, but I have not tested; and I am not sure how to test it.
> Documentation for NODE_OPTIONS are here:
> https://nodejs.org/docs/latest-v12.x/api/cli.html#cli_node_options_options

Its configured in the project thus one can always create a `.env` file to pass the environment variables

File: openshift/README.md
Comment (mrmodise) [inline openshift/README.md:16]: is this password protected?

Comment (mrmodise) [reply openshift/README.md:16]: Similarly, for Redis how does the password configuration work?

**Comment by mrmodise** [comment]
@johanschvn alternative you can edit the `README.md` and highlight that this setup should not be used in production as-is without further security and performance optimisation configurations.

With that comment am happy to ignore the above and land the PR

**Comment by johanschvn** [comment]
> @johanschvn alternative you can edit the `README.md` and highlight that this setup should not be used in production as-is without further security and performance optimisation considerations.
> 
> With that comment am happy to ignore the above and land the PR

@mrmodise thank you for your feedback
I agree that this is not an instruction for how to deploy to production.
I will add something about that this is a quick and easy way to get the example deployed into OpenShift for demo or test purpose and not production. I have put the PR in draft mode until ready for new review.
I am considering also adding more info:
* saying something about persistence - with this setup the databases are not persisted - if pod is moved to an other node, data will be lost
* maybe add instruction on how to expose the application as https instead of http  
* maybe add instructions on how to set resource limits
* maybe explaining why there are specific Dockerfile.* for building images for OpenShift

File: openshift/recommender.yaml
Comment (johanschvn) [reply openshift/recommender.yaml:14]: This should be in DeploymentConfig not in BuildConfig. I am considering adding info on how to do that.

File: openshift/README.md
Comment (johanschvn) [reply openshift/README.md:16]: There are no password protections for the databases.
The datasources expects that the databases are not password protected.
This is not something that is OpenShift specific, this is the current state of the application.
To start using passwords you have to change mango.datasource.ts and redis.datasource.ts.

The "--name shopping-app-mongodb" is for using the same solution as for Kubernetes example.

File: openshift/recommender.yaml
Comment (johanschvn) [reply openshift/recommender.yaml:14]: Added info about setting resource constraints to README.md in the current PR (force pushed with history rewritten).

**Review by mrmodise — APPROVED** [review]

**Comment by mrmodise** [comment]
LGTM @johanschvn thanks for the PR as always. We can wait for one more approval then merge. 

PTAL @hacksparrow @jannyHou

**Comment by mrmodise** [comment]
Closes #935

File: openshift/README.md
Comment (hacksparrow) [inline openshift/README.md:1]: A link to this doc from https://github.com/strongloop/loopback4-example-shopping/blob/master/README.md would be helpful. Please create an appropriately named section (inspiration https://github.com/strongloop/loopback4-example-shopping/blob/master/README.md#deploy-to-cloud-as-microservices).

@johanschvn @mrmodise

Comment (johanschvn) [reply openshift/README.md:1]: @hacksparrow @mrmodise what do you think about adding this after the Deploy to Cloud as Microservices section: 
```
## Build and deploy on Red Hat OpenShift

You can find [instructions](openshift/README.md), Dockerfiles and resource
definition files for building and deploying on Red Hat OpenShift Container
Platform in the openshift directory.
```

Comment (hacksparrow) [reply openshift/README.md:1]: LGTM

**Comment by hacksparrow** [comment]
Thank you @johanschvn!

**Pull request merged**
